py/server/deephaven/experimental/table_data_service.py

```diff
@@ -0,0 +1,345 @@
+            else:
+                if pt_table.num_rows != 1:
+                    raise ValueError("The number of rows in the pyarrow table for partition column values must be 1")
+                bb_list = [jpy.byte_buffer(rb.serialize()) for rb in pt_table.to_batches()]
+                bb_list.insert(0, jpy.byte_buffer(pt_table.schema.serialize()))
+                callback.accept(j_tbl_location_key, jpy.array("java.nio.ByteBuffer", bb_list))
+
+        return self._backend.subscribe_to_table_locations(table_key, callback_proxy)
+
+    def _table_location_size(self, table_key: TableKey, table_location_key: TableLocationKey, callback: jpy.JType):
+        """ Provides the size of the table location with the given table key and table location key to the table service
+        in the engine.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (TableLocationKey): the table location key
+            callback (jpy.JType): the Java callback function with one argument: the size of the table location in number
+                of rows
+        """
+        def callback_proxy(size):
+            callback.accept(size)
+
+        self._backend.table_location_size(table_key, table_location_key, callback_proxy)
+
+    def _subscribe_to_table_location_size(self, table_key: TableKey, table_location_key: TableLocationKey,
+                                             callback: jpy.JType) -> Callable[[], None]:
+        """ Provides the current and future sizes of the table location with the given table key and table location key
+        to the table service in the engine.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (TableLocationKey): the table location key
+            callback (jpy.JType): the Java callback function with one argument: the size of the partition in number of
+                rows
+
+        Returns:
+            Callable[[], None]: a function that can be called to unsubscribe from this subscription
+        """
+        def callback_proxy(size):
+            callback.accept(size)
+
+        return self._backend.subscribe_to_table_location_size(table_key, table_location_key, callback_proxy)
+
+    def _column_values(self, table_key: TableKey, table_location_key: TableLocationKey, col: str, offset: int,
+                       min_rows: int, max_rows: int, callback: jpy.JType) -> None:
+        """ Provides the data values for the column with the given name for the table location with the given table key
+        and table location key to the table service in the engine.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (TableLocationKey): the table location key
+            col (str): the column name
+            offset (int): the starting row index
+            min_rows (int): the minimum number of rows to return
```

Comment:
Will always be <= page_size. We can guarantee that for min, we shouldn't over-specify max, I think.

Reply:
What do you mean by shouldn't over specify max? I've added the comment about min_rows. Right now max_rows is keyed off of the chunk's capacity. Is it possible that this capacity is over-specified?